### PR TITLE
[persist/sources] Choose a resumption frontier in the source based on the output upper, not the input since

### PR DESCRIPTION
**Pull request by bkirwi**
- Choose the "as-of" for the source as a function of the output uppers, not the input since. (While the since is normally in advance of the uppers, this does not appear to be the case when the source is being created or shutting down.)
  - This is morally similar to how the controller places advances its read holds on upstream collections based on the progress of the upper in downstream collections.
- When we open a read handle for a shard, start from whatever since is available. (And take care to ensure that we don't expose any insufficiently-advanced data to downstreams.)
 
### Motivation

For: #26940.

The previous code depends on brittle details of the way that the since is chosen, including leaking a handle briefly to temporarily prevent compaction. The result is... still brittle, frankly; but it's hopefully a step towards a less brittle future where the since of Persist shards is easier to reason about.

### Tips for reviewer

Nightly here: https://buildkite.com/materialize/nightly/builds/7922

This may still be a little janky, but I'd figured I'd better get it up and let the source experts tell me exactly which parts are the janky ones.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/storage/src/source/reclock/compat.rs [hunk 1]
```diff
@@ -102,23 +103,11 @@ where
...
-        // them offer the invariant that we need (that the `as_of` if <= this
-        // `since`). Using a `SinceHandle` here does not offer the same
-        // invariant when paired with a `ReadHandle`.
-        let since = read_handle.since();
+        let since = read_handle.since().clone();
 
         // Allow manually simulating the scenario where the since of the remap
         // shard has advanced too far.
         fail_point!("invalid_remap_as_of");
-        assert!(
```
Comment (aljoscha) [inline src/storage/src/source/reclock/compat.rs:115]: Where are we now doing something that is equivalent to this assert? When/if this doesn't hold, we might produce duplicate updates to the source shard, afaik.

File: src/storage/src/source/reclock/compat.rs [hunk 2]
```diff
@@ -135,13 +124,13 @@ where
         use futures::stream;
         let events = stream::once(async move {
             let updates = read_handle
-                .snapshot_and_fetch(as_of.clone())
```
Comment (petrosagg) [inline src/storage/src/source/reclock/compat.rs:138]: Why are we not using the `as_of` here since we are advancing the timestamps to that frontier below anyway?

File: src/storage/src/source/reclock/compat.rs [hunk 3]
```diff
@@ -154,6 +143,18 @@ where
...
         .flatten()
+        .map(move |mut event| {
+            // NB: we advance the updates to the as-of frontier, to ensure that downstream code
+            // doesn't observe distinctions between times it's not meant to care about.
+            if let ListenEvent::Updates(updates) = &mut event {
+                for (_, t, _) in updates.iter_mut() {
+                    t.advance_by(as_of.borrow())
+                }
+                // This is not necessary, but we might as well save downstream code some effort.
+                consolidate_updates(updates);
```
Comment (petrosagg) [inline src/storage/src/source/reclock/compat.rs:154]: The consuming code also consolidates so this is call can be removed

File: src/storage/src/source/reclock/compat.rs (hunk 1 above)
Comment (petrosagg) [reply src/storage/src/source/reclock/compat.rs:115]: I'm curious about this removal too

Comment (bkirwi) [reply src/storage/src/source/reclock/compat.rs:115]: Fair! In this PR we are worried about:

> While the since is normally in advance of the uppers, this does not appear to be the case when the source is being created or shutting down.

If we want to choose an as-of based on the uppers of the shard and not the since of the shard, the condition this assert checks for will become normal. So this PR makes some effort to make sure it is also okay. For example, in the startup case the global since of the shard is way ahead of the upper, and thus the as-of... but we happen to know that no data will be produced in that interval, so reading from the since and accumulating to the as-of will produce the same result.

(Ideally the since would be held not past the upper in the controller so the shard was always readable, but that's a separate can of worms...)

File: src/storage/src/source/reclock/compat.rs (hunk 2 above)
Comment (bkirwi) [reply src/storage/src/source/reclock/compat.rs:138]: Same concern as here, basically: https://github.com/MaterializeInc/materialize/pull/27314#discussion_r1620860314

Another possibly-more-readable option would be to start reading from the greater of the two frontiers.

File: src/storage/src/source/reclock/compat.rs (hunk 1 above)
Comment (bkirwi) [reply src/storage/src/source/reclock/compat.rs:115]: I pushed a commit that re-adds the assert with a carveout for `[0]`. Would that sort of thing be better?

File: src/storage/src/source/reclock/compat.rs (hunk 2 above)
Comment (bkirwi) [reply src/storage/src/source/reclock/compat.rs:138]: (Pushed a commit with that idea; LMK what you think.)

Comment (petrosagg) [reply src/storage/src/source/reclock/compat.rs:138]: Ah, I see. That makes sense but I don't think we should have this logic here. The `as_of` parameter basically communicates "the collection accumulates correctly as of this frontier" but that's not true in the cases you mentioned. This leads me to the conclusion that the `as_of` should never have had this value in the first place. My suggestion would be to change the async storage worker to set as_of to `(upper - 1).join(since)`

Comment (bkirwi) [reply src/storage/src/source/reclock/compat.rs:138]: > The as_of parameter basically communicates "the collection accumulates correctly as of this frontier" but that's not true in the cases you mentioned.

I think it might be true! Here's the reasoning:
- In normal operation, the new as-of should be past the since, and so if it accumulated correctly at the since it should accumulate correctly at the as-of.
- On start-up, the normal relationship is inverted... the as-of is at `[0]` and the since is at some greater frontier `F > [0]`. Normally, we'd have no reason to think that accumulating at `[0]` would give the same result as accumulating at `F`. But _in the case of the source specifically_, we happen to know that the collection will be empty between `[0` and `[F]`, since we don't reclock data before our initial since, so the fact that we start reading later doesn't matter.
    - It seems gross taking advantage of this special property, but note that the source is _already_ taking advantage of this special property... if this wasn't true, and we could end up reclocking input data to some time before `F` in our source exports, then accumulating up to `F` would give bad results.

Comment (petrosagg) [reply src/storage/src/source/reclock/compat.rs:138]: > On start-up, the normal relationship is inverted... the as-of is at `[0]` and the since is at some greater frontier `F > [0]`

This is what doesn't sound right to me. Each collection has an associated `since` frontier, which defines the minimum times beyond which you can accumulate to correct results. When a dataflow is rendered that computes a result based on some input we can define the set of times that we're interested in for the result. This is what an `as_of` frontier communicates and it only makes sense when it's greater than or equal to the since frontier. It basically says "I know you have all this data since the `since` frontier, but I only want to compute the states as of the `as_of` frontier, so please advance all the times and compact them so I do less work".

> But in the case of the source specifically, we happen to know that the collection will be empty between [0 and [F],

It is undefined to talk about the contents of a collection for times that are not beyond its since frontier so I'm not sure what it means that we happen to know its contents before the since frontier F.

That said, the remap collection couldn't possibly be empty between 0 and F, if that was the case then that would mean that at timestamp 0 the remap collection would accumulate to the empty frontier, which would mean that the entirety of the source data would need to be reclocked to that time. But again, this is a bit meaningless when the since frontier disallows talking about those times.

Is something broken if we change the async storage worker to say "we want to read the remap collection as of its since frontier or upper - 1, whichever is greater"?

Comment (bkirwi) [reply src/storage/src/source/reclock/compat.rs:138]: > That said, the remap collection couldn't possibly be empty between 0 and F, if that was the case then that would mean that at timestamp 0 the remap collection would accumulate to the empty frontier, which would mean that the entirety of the source data would need to be reclocked to that time.

Good callout, thank you!

> But again, this is a bit meaningless when the since frontier disallows talking about those times.

Agree that it's not particularly meaningful to talk about the contents of the _shard_, but I think it's fair to talk about the contents of the "logical collection", as you've just shown! 😅

Anyways - I'll see if I can do one more take on this, based only on "stable" information about the outputs, and report back.

Comment (bkirwi) [reply src/storage/src/source/reclock/compat.rs:138]: Okay, I think the latest version maybe nobody will hate!

The idea here is to use the since of the _export_ collection(s), not the remap collection. The general idea is that we if nobody will ever read any of our exports as-of some time, we don't need to reproduce our export as-of that time, and thus we don't need to read the remap collection as-of that time. This helps in the specific problem case on bootstrap, where the controller had advanced the since well past the upper... but it also doesn't rely on any particular ordering of downgrades in the controller, which solves my problem.

Comment (petrosagg) [reply src/storage/src/source/reclock/compat.rs:138]: > The general idea is that we if nobody will ever read any of our exports as-of some time, we don't need to reproduce our export as-of that time, and thus we don't need to read the remap collection as-of that time. 

That would indeed the most correct statement! However, where do we ensure that by the time this async worker reads the since frontier of an export it is not still at `{0}`? There have been many refactors in the controller and I couldn't trace where the since handle is used. Is the controller ensuring that shard initialization has finished before sending the command to the cluster?

File: src/storage/src/source/reclock/compat.rs [hunk 4]
```diff
@@ -112,49 +112,60 @@ where
...
-
-            let listener = read_handle
-                .listen(as_of.clone())
-                .await
-                .expect("since <= as_of asserted");
-
-            let listen_stream = stream::unfold(listener, |mut listener| async move {
-                let events = stream::iter(listener.fetch_next().await);
-                Some((events, listener))
+        let events = if since.is_empty() {
```
Comment (petrosagg) [inline src/storage/src/source/reclock/compat.rs:120]: Is this change needed? Does the new since selection logic trigger this case more often?

File: src/storage/src/source/reclock/compat.rs (hunk 2 above)
Comment (bkirwi) [reply src/storage/src/source/reclock/compat.rs:138]: Two answers:
- [`StorageController::create_collections`](https://github.com/MaterializeInc/materialize/blob/52ee3f0fe46d23869274d6d7d8b47aaf48a7aab0/src/storage-client/src/storage_collections.rs#L1296) appears to downgrade the since handles as part of `open_data_handles`, which should occur before the actual exports are created. Though if @aljoscha can confirm this that is helpful!
- If the since was only advanced asynchronously, I believe we'd already have problems: since any reader that could read the shard before that initial since advancement could hold back the since, possibly reading data as of a time before the source was created and thus before we could possibly have a correct result. And if so this property is already required for correctness...

File: src/storage/src/source/reclock/compat.rs (hunk 4 above)
Comment (bkirwi) [reply src/storage/src/source/reclock/compat.rs:120]: Yup! For example, an ill-timed drop could cause the since to advance in between the time the as-of was chosen and this code. (This was theoretically possible in the old code too, but it would have needed to take >5 minutes between the time the as-of was chosen and the time the dataflow actually started, which I've never seen.)

IIUC it's also possible for the remap shard to be finalized before the exports/subsources, though I don't understand that code well enough to know for sure...

File: src/storage/src/source/reclock/compat.rs (hunk 2 above)
Comment (petrosagg) [reply src/storage/src/source/reclock/compat.rs:138]: I think what I may be remembering is the time before Sean fixed the subsource dependency inversion thing where the controller would first learn about subsources (with no context on which source they belong to) and only afterwards learn about the source itself.

> And if so this property is already required for correctness...

I think correctness property #2 is not working for `ADD SUBSOURCE` for a bunch of reasons that need.. (you guessed it) some of the controller refactors to land.

Cool, then the current state looks good to me!

**Review by petrosagg — APPROVED** [review]

File: src/storage/src/source/reclock/compat.rs (hunk 2 above)
Comment (bkirwi) [reply src/storage/src/source/reclock/compat.rs:138]: Thanks for working through it with me! A bit of a winding road but I'm pretty happy with where we ended up.

I'll get this in to let it bake a bit before the release cut this week...
